From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390: do not bypass BPENTER for interrupt system calls
Patch-mainline: v4.16-rc5
Git-commit: d5feec04fe578c8dbd9e2e1439afc2f0af761ed4
References: bnc#1089386, LTC#166572

Description:  kernel: expoline defense against the spectre attack
Symptom:      None
Problem:      The spectre attack may be used to read restricted kernel
              data from user space.
Solution:     Provide execute trampolines for indirect branches
Reproduction: None

Upstream-Description:

              s390: do not bypass BPENTER for interrupt system calls

              The system call path can be interrupted before the switch back to the
              standard branch prediction with BPENTER has been done. The critical
              section cleanup code skips forward to .Lsysc_do_svc and bypasses the
              BPENTER. In this case the kernel and all subsequent code will run with
              the limited branch prediction.

              Fixes: eacf67eb9b32 ("s390: run user space and KVM guests with modified branch prediction")
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/kernel/entry64.S |    1 +
 1 file changed, 1 insertion(+)

--- a/arch/s390/kernel/entry64.S
+++ b/arch/s390/kernel/entry64.S
@@ -1233,6 +1233,7 @@ cleanup_update:
 	jz	0f
 	stg	%r11,__TI_last_break(%r12)
 0:	mvc	__LC_RETURN_PSW+8(8),BASED(cleanup_table_system_call+8)
+	BPENTER __TI_flags+6(%r12),_TIF_ISOLATE_BP>>8
 	la	%r12,__LC_RETURN_PSW
 	BR_R13USE_R14
 cleanup_system_call_insn:
